### PR TITLE
menubar: consider XDG_HOME_DIR and XDG_DATA_DIRS

### PR DESCRIPTION
**Pull request by romildo**
The freedesktop specifications let desktop files to be stored in
different directories indicated by the environment variables
XDG_DATA_HOME and XDG_DATA_DIRS. Only use the default value for these
variables if the variables are not defined.

This is important for systmes like NixOS which does not follow the LFS
and installs files differently.

### CODE REVIEW

**Comment by Veratil** [comment]
Your `split` and `map` functions could be useful outside of this, along with determining the `xdg_data_*` variables.

For modules in PR waiting:
`split` could move to `gears.string`
`map` could move to `gears.table`
`xdg_data_home` and `xdg_data_dirs` could move to functions in `gears.filesystem`

Then add the few requires() and this would be as simple as:
```lua
local gtable = require("gears.table")
local gstring = require("gears.string")
local gfs = require("gears.filesystem")
menu_gen.all_menu_dirs = gtable.map (
    function(dir) return dir .. '/applications/' end,
    gstring.split(gfs.get_xdg_data_home() .. ':' .. gfs.get_xdg_data_dirs(), ':'))
```
with the added benefit of everybody else getting to use it if they need. :)

**Comment by Veratil** [comment]
@romildo `gears.string` and `gears.table` have been merged if you want to add those functions to their respective modules. `gears.filesystem` is approved and going to be merged soon, too.

Don't forget the documentation! :grin:

**Comment by romildo** [comment]
@Veratil I have reworked the PR so that the helper functions are defined in the appropriate module in the `gears` library.

**Comment by Elv13** [comment]
The CI found 1 problem

```
Checking lib/gears/string.lua                     1 warning

    lib/gears/string.lua:89:38: trailing whitespace in a comment
```

File: lib/gears/table.lua [hunk 1]
```diff
@@ -212,5 +212,20 @@ function gtable.merge(t, set)
...
 end
 
+--- Map a function to a table. The function is applied to each value
+-- on the table, returning a table of the results.
+-- @class function
+-- @name map
+-- @tparam function f the function to be applied to each value on the table
+-- @tparam table tbl the container table whose values will be operated on
+-- @treturn table Return a table of the results
+function gtable.map(f, tbl)
```
Comment (Elv13) [inline lib/gears/table.lua]: Could you swap both parameters for consistency with the other function?

Comment (Veratil) [inline lib/gears/table.lua]: In defense of the argument order, this is traditionally the way the map function arguments are in all languages.
Python: `map(function_to_apply, list_of_inputs)`
Haskell: `map :: (a -> b) -> [a] -> [b]`
PHP: `array array_map ( callable $callback , array $array1 [, array $... ] )`
Even the [Lua example](http://lua-users.org/wiki/FunctionalLibrary) has `(func, tbl)`

Comment (Elv13) [inline lib/gears/table.lua]: ok, then

**Comment by Elv13** [comment]
Merged

**Pull request closed without merging**